numpyro/distributions/util.py

```diff
@@ -13,21 +13,149 @@
 from jax.util import partial
 
 
-@partial(jit, static_argnums=(3, 4))
-def _binomial(key, p, n, n_max, shape):
+_tr_params = namedtuple('tr_params', ['c', 'b', 'a', 'alpha', 'u_r', 'v_r', 'm', 'log_p', 'log1_p', 'log_h'])
+
+
+def _get_tr_params(n, p):
+    mu = n * p
+    spq = np.sqrt(mu * (1 - p))
+    c = mu + 0.5
+    b = 1.15 + 2.53 * spq
+    a = -0.0873 + 0.0248 * b + 0.01 * p
+    alpha = (2.83 + 5.1 / b) * spq
+    u_r = 0.43
+    v_r = 0.92 - 4.2 / b
+    m = np.floor((n + 1) * p).astype(n.dtype)
+    log_p = np.log(p)
+    log1_p = np.log1p(-p)
+    log_h = (m + 0.5) * np.log((m + 1.) / (n - m + 1.)) + log1_p - log_p + \
+            (stirling_approx_tail(m) + stirling_approx_tail(n - m))
+    return _tr_params(c, b, a, alpha, u_r, v_r, m, log_p, log1_p, log_h)
+
+
+def stirling_approx_tail(k):
+    precomputed = [
+        0.08106146679532726,
+        0.04134069595540929,
+        0.02767792568499834,
+        0.02079067210376509,
+        0.01664469118982119,
+        0.01387612882307075,
+        0.01189670994589177,
+        0.01041126526197209,
+        0.009255462182712733,
+        0.008330563433362871,
+    ]
+    kp1 = k + 1
+    kp1sq = (k + 1) ** 2
+    return np.where(k < 10,
+                    np.take(precomputed, k),
+                    (1. / 12 - (1. / 360 - (1. / 1260) / kp1sq) / kp1sq) / kp1)
+
+
+def _binomial_btrs(key, p, n):
+    """
+    Based on the transformed rejection sampling algorithm (BTRS) from the
+    following reference:
+
+    Hormann, "The Generation of Binonmial Random Variates"
+    (https://core.ac.uk/download/pdf/11007254.pdf)
+    """
+    def _btrs_body_fn(val):
+        _, key, _, _ = val
+        key, key_u, key_v = random.split(key, 3)
+        u = random.uniform(key_u)
+        v = random.uniform(key_v)
+        u = u - 0.5
+        k = np.floor((2 * tr_params.a / (0.5 - np.abs(u)) + tr_params.b) * u + tr_params.c).astype(n.dtype)
+        return k, key, u, v
+
+    def _btrs_cond_fn(val):
+        def accept_fn(k, u, v):
+            m = tr_params.m
+            log_p = tr_params.log_p
+            log1_p = tr_params.log1_p
+            log_f = (n + 1.) * np.log((n - m + 1.) / (n - k + 1.)) + \
+                    (k + 0.5) * np.log((n - k + 1.) / (k + 1.)) + log_p - log1_p + \
+                    (stirling_approx_tail(k) - stirling_approx_tail(n - k)) + tr_params.log_h
+            log_g = (tr_params.a / (0.5 - np.abs(u)) ** 2) + tr_params.b
+            return np.log(v) <= log_f + log_g - np.log(tr_params.alpha)
```

Comment:
I guess you want to compute `np.log(v * tr_params.alpha / g) <= log_f` here.

Reply:
This should be fixed; also added a comment.

Reply:
I still think using 1 `log` as above is better than computing 3 `log` terms here.